### ISSUE TITLE
Add more options for encryption

### ISSUE THREAD

**Issue by MuntashirAkon** [MuntashirAkon/AppManager#116]
Currently AM only allows encryption through OpenPGP clients (this is still the most safe and recommended method). However, support for other encryptions are useful for many users (despite various drawbacks). The crypto utils in AM are designed to support more encryption methods. It is also preferable that user generate new keys (as well as import them). But I'm yet to investigate on how many indices to add in order to generate keys since I intend to support both symmetric and asymmetric encryption methods with variable key size.

Import/export feature seems obvious but import/exporting keys in plain text (like what Bitwarden does) is unsafe. I'm thinking of using the import/export mechanism used by Signal and OpenKeychain.

**Comment by MuntashirAkon** [comment]
It seems that simply encryption type and size is enough to generate a keys.

We'll be using AES with mode GCM, RSA and possibly EC with Android KeyStore as the secure random provider to generate 128/192/256 bits keys (not sure whether Android allows AES 192 or 256 though since standard JDK/JRE requires a jurisdiction file).

**Matrix implementation:**
https://github.com/matrix-org/matrix-android-sdk/blob/3d3c42c5a55bcc4e80d268a6eace93eb1dfc0247/matrix-sdk-core/src/main/java/org/matrix/androidsdk/core/CompatUtil.java#L107

**Signal implementation:**
https://github.com/signalapp/Signal-Android/blob/e75a03b6f884dbc38eb8173e97b673b24c407ac0/app/src/main/java/org/thoughtcrime/securesms/crypto/KeyStoreHelper.java#L86

**Comment by MuntashirAkon** [comment]
> It seems that simply encryption type and size is enough to generate a keys.
> 
> We'll be using AES with mode GCM, RSA and possibly EC with Android KeyStore as the secure random provider to generate 128/192/256 bits key (not sure whether Android allows AES 192 or 256 though since standard JDK/JRE requires a jurisdiction file) for AES, 1024/2048/4096 bits key for RSA and so on.
> 
> **Matrix implementation:**
> https://github.com/matrix-org/matrix-android-sdk/blob/3d3c42c5a55bcc4e80d268a6eace93eb1dfc0247/matrix-sdk-core/src/main/java/org/matrix/androidsdk/core/CompatUtil.java#L107
> 
> **Signal implementation:**
> https://github.com/signalapp/Signal-Android/blob/e75a03b6f884dbc38eb8173e97b673b24c407ac0/app/src/main/java/org/thoughtcrime/securesms/crypto/KeyStoreHelper.java#L86

**Comment by MuntashirAkon** [comment]
### Key generation
In Settings, **OpenPGP Provider** is replaced by **Encryption**. This contains three options: **AES**, **RSA** and **OpenPGP** (we will not be working on ECC for now because it may require a hybrid encryption scheme such as [this](https://cryptobook.nakov.com/asymmetric-key-ciphers/ecc-encryption-decryption)).

**AES:** AES has a separate dialog where user can generate or input existing keys. In case of key generation, there is an option to choose key size among 128/192/256.

**RSA:** _(and in future, ECC)_ Asymmetric cryptography requires a private and a public key. In its dedicated dialog, there is a option to generate the key pair or import existing ones. In case of key generation, there is an option to choose key size among 1024/2048/4096.

### Changes in the meta
`mode` doesn't seem to be the right choice to store encryption types. Therefore, this will be replaced by `crypto` which contains one of the following: `none`, `aes`, `rsa` or `pgp`.

For `aes` and `rsa`, we need another field namely `iv` (or nonce) which is needed to ensure that each backup is unique. This applies to individual backup, not files.

For `rsa`, public keys needed to be stored for individual backups as well.

For `pgp`, the key ids have to be saved for individual backup, so that the backups will not be lost in case the user forgot which key(s) they used. For this, a field `key_ids` is needed to be added.

**Comment by MuntashirAkon** [comment]
### Storing keys
Like key generation, the initial plan was to store keys at Android KeyStore but since it allows exporting keys _only_ from API 28. It's not possible to store keys there as it's compulsory for AM to export keys from API 21. Therefore, the solution is to use BountyCastle. The only known app that use BC directly on Android is [OpenKeychain](https://github.com/open-keychain/open-keychain).

We'll also consider using one of the BC KeyStore instead of the default Java KeyStore (JKS). But it depends on #138. To be more precise, we'll be using a single KeyStore to save both custom certificate info and encryption keys.

**Comment by MuntashirAkon** [comment]
After considering a lot of ways, I've concluded that the following is the most safest way to store keys without additional overhead:

AM will maintain a single KeyStore (as described earlier) which will be password protected (the user will be offered to save the password to some place else). The keys will also be protected using password (same password as the KeyStore if the user doesn't choose to insert a new password). Finally, the passwords will be saved in the shared preferences but not in the plain text. They will be encrypted using RSA encryption which will be stored in Android KeyStore. Now, if the user decides to take a backup of AM, the KeyStore will not work. In that case, AM will ask to insert the passwords again. This has other benefits such as exporting the KeyStore will be very easy and won't be insecure since the KeyStore is protected with a password only known to the user.

**Caveat:** Since users can add multiple keys, it might be necessary to make an actual key management system which brings the overhead that I talked about in #30.
